OpenSim/Simulation/Model/Smith2016Ligament.cpp

```diff
@@ -0,0 +1,377 @@
+
+
+
+
+void Smith2016Ligament::setNull()
+{
+	setAuthors("Colin Smith");
+	setReferences(
+	"Smith, C.R., Lenhart, R.L., Kaiser, J., Vignos, M.F. and Thelen, D.G.,"
+	"(2016). Influence of ligament properties on tibiofemoral mechanics"
+	" in walking. J Knee Surg, 29(02), 99-106.\n\n"
+	
+	"Wismans, J.A.C., Veldpaus, F., Janssen, J., Huson, A. and Struben, P.," 
+	"(1980). A three-dimensional mathematical model of the knee-joint. "
+	"J Biomech, 13(8), 677-685.\n\n"
+
+	"Blankevoort, L. and Huiskes, R., (1991)."
+	"Ligament-bone interaction in a three-dimensional model of the knee."
+	"J Biomech Eng, 113(3), 263-269"
+	);
+}
+
+void Smith2016Ligament::constructProperties()
+{
+	constructProperty_GeometryPath(GeometryPath());
+	constructProperty_linear_stiffness(0.0);
+	constructProperty_transition_strain(0.06);
+	constructProperty_normalized_damping_coefficient(0.003);
+	constructProperty_slack_length(0.0);
+	
+}
+
+void Smith2016Ligament::extendFinalizeFromProperties()
+{
+	Super::extendFinalizeFromProperties();
+
+	//Add channels to output list
+	auto& dyn_quan = updOutput("dynamic_quantities");
+	dyn_quan.addChannel("force_spring");
+	dyn_quan.addChannel("force_damping");
+	dyn_quan.addChannel("force_total");
+	dyn_quan.addChannel("length");
+	dyn_quan.addChannel("lengthening_speed");
+	dyn_quan.addChannel("strain");
+	dyn_quan.addChannel("strain_rate");	
+
+
+	//Check that properties are valid
+	OPENSIM_THROW_IF_FRMOBJ(get_slack_length() < 0.,
+        InvalidPropertyValue, getProperty_slack_length().getName(),
+        "Slack Length cannot be less than 0");
+
+	OPENSIM_THROW_IF_FRMOBJ(get_linear_stiffness() < 0.,
+        InvalidPropertyValue, getProperty_linear_stiffness().getName(),
+        "Linear Stiffness cannot be less than 0");
+
+	OPENSIM_THROW_IF_FRMOBJ(get_normalized_damping_coefficient() < 0.,
+        InvalidPropertyValue, getProperty_normalized_damping_coefficient().getName(),
+        "Normalized Damping Coefficient cannot be less than 0");
+
+	OPENSIM_THROW_IF_FRMOBJ(get_transition_strain() < 0.,
+        InvalidPropertyValue, getProperty_transition_strain().getName(),
+        "Transistion Strain cannot be less than 0");
+
+	//Set Ligament Color
+	GeometryPath& path = upd_GeometryPath();
+    path.setDefaultColor(SimTK::Vec3(0.99, 0.42, 0.01));
+
+}
+
+void Smith2016Ligament::extendAddToSystem(SimTK::MultibodySystem& system) const
+{
+	Super::extendAddToSystem(system);
+
+	addCacheVariable<double>("force_spring",0.0, SimTK::Stage::Dynamics);
+	addCacheVariable<double>("force_damping", 0.0, SimTK::Stage::Dynamics);
+	addCacheVariable<double>("force_total", 0.0, SimTK::Stage::Dynamics);
+	addCacheVariable<double>("length", 0.0, SimTK::Stage::Dynamics);
+	addCacheVariable<double>("lengthening_speed", 0.0, SimTK::Stage::Dynamics);
+	addCacheVariable<double>("strain", 0.0, SimTK::Stage::Dynamics);
+	addCacheVariable<double>("strain_rate", 0.0, SimTK::Stage::Dynamics);
```

Comment:
These cache variables either depend on Position or Velocity, not Dynamics. Even the force only depends on the Velocity stage.